### PR TITLE
Rotation With Quaternion

### LINKED ISSUE #9510: Rotation representation with quaternion

### PR DESCRIPTION
**Pull request by super626**
fix #9510

### CODE REVIEW

File: cocos/2d/CCNode.h [hunk 1]
```diff
@@ -1614,6 +1630,8 @@ class CC_DLL Node : public Ref
     // rotation Z is decomposed in 2 to simulate Skew for Flash animations
     float _rotationZ_X;             ///< rotation angle on Z-axis, component X
     float _rotationZ_Y;             ///< rotation angle on Z-axis, component Y
+    
+    Quaternion _rotation_quat;      ///rotation using quaternion, if _rotationZ_X == _rotationZ_Y, _rotation_quat = RotationZ_X * RotationY * RotationX, else _rotation_quat = RotationY * RotationX
```
Comment (minggo) [inline cocos/2d/CCNode.h]: _rotation_quat -> _rotationQuat.

Please remember it.
Thanks.

**Comment by minggo** [comment]
@ricardoquesada 
@dabingnn 
@darkdukey 
Could you please help to review it.
It is an important modification.

Thanks.

File: cocos/2d/CCNode.h (hunk 1 above)
Comment (super626) [inline cocos/2d/CCNode.h]: Sorry, got it.
I see the variable above is _rotationZ_Y, but it is different.

File: cocos/2d/CCNode.cpp [hunk 2]
```diff
@@ -374,6 +378,44 @@ Vec3 Node::getRotation3D() const
     return Vec3(_rotationX,_rotationY,_rotationZ_X);
 }
 
+void Node::updateRotationQuat()
+{
+    // convert Euler angle to quaternion
+    // when _rotationZ_X == _rotationZ_Y, _rotation_quat = RotationZ_X * RotationY * RotationX
+    // when _rotationZ_X != _rotationZ_Y, _rotation_quat = RotationY * RotationX
+    float halfRadx = CC_DEGREES_TO_RADIANS(_rotationX / 2.f), halfRady = CC_DEGREES_TO_RADIANS(_rotationY / 2.f), halfRadz = _rotationZ_X == _rotationZ_Y ? -CC_DEGREES_TO_RADIANS(_rotationZ_X / 2.f) : 0;
+    float coshalfRadx = cosf(halfRadx), sinhalfRadx = sinf(halfRady), coshalfRady = cosf(halfRady), sinhalfRady = sinf(halfRady), coshalfRadz = cosf(halfRadz), sinhalfRadz = sinf(halfRadz);
```
Comment (dabingnn) [inline cocos/2d/CCNode.cpp]: Could you please change the variables to multiple lines? It is hard to read and review.

File: cocos/2d/CCNode.cpp [hunk 3]
```diff
@@ -1708,11 +1754,12 @@ const Mat4& Node::getNodeToParentTransform() const
             y += _anchorPointInPoints.y;
         }
 
+        bool needsSkewMatrix = ( _skewX || _skewY );
```
Comment (dabingnn) [inline cocos/2d/CCNode.cpp:1757]: Is the new code of `Node::getNodeToParentTransform` compatible with the old one?

Comment (super626) [inline cocos/2d/CCNode.cpp:1757]: Sure.

Comment (dabingnn) [inline cocos/2d/CCNode.cpp:1757]: Thanks

**Comment by dabingnn** [comment]
The pull request changes the logic of `getNodetoParentTransform`, which may be called frequently if the node contains actions.
It is better to add a performance test result.

**Comment by super626** [comment]
I had thought about this, too.
We will share a performance test result.

**Comment by super626** [comment]
[ci rebuild]

**Comment by minggo** [comment]
Could you please add test case for it?

**Comment by super626** [comment]
Ok, I will

**Pull request merged**